core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -677,8 +677,8 @@ public SqlNode toSql(RexProgram program, RexNode rex) {
         }
         if (op instanceof SqlBinaryOperator && nodeList.size() > 2) {
           // In RexNode trees, OR and AND have any number of children;
-          // SqlCall requires exactly 2. So, convert to a left-deep binary tree.
-          return createLeftCall(op, nodeList);
+          // SqlCall requires exactly 2. So, convert to a balanced binary tree.
+          return createBinaryCall(op, nodeList, 0, nodeList.size());
```

Comment:
Not sure if I should apply balanced binary tree to AND/OR operators only, for operators that are not associative, we may still need left-deep tree. But is there any other operator that can have more than 2 children and not associative?